### ISSUE TITLE
Validation losses are not computed properly: they're stuck to 0 or "inf" 

### ISSUE THREAD

**Issue by robertastellino** [ultralytics/ultralytics#4785]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, I trained a Yolov8 model for object detection on my custom dataset. The training went fine, but looking at the results (which I have attached below), I saw that the plots regarding the training are just as I expected, while the plots regarding the validation are suspiscious: the validation loss and the val/dfl loss are stuck to 0 and the val/cls loss is "inf". What could be the problem? 

![results](https://github.com/ultralytics/ultralytics/assets/138810693/d265e422-2dde-4fce-bdc4-8987656a3a20)

The data I presented to the network are organized as follows:

```
images
  ├── train
  ├── test
  ├── val

labels 
  ├── train
  ├── test
  ├── val
```

and my data.yaml file contains the following:

```
path: '/Home/Dati/Data/' # dataset root dir
train:  images/train # train images (relative to 'path')
val:    images/val # val images (relative to 'path')
test:   images/test

# Classes
names:
  0: plane
  1: car

```
If there is any information you need to know about the training that I didn't mentioned above, don't hesitate to ask, I am obviously available to answer.
Thank you for your attention.

**Comment by glenn-jocher** [comment]
@robertastellino hello! Thank you for reaching out with your question concerning validation losses that are not being computed properly in your YOLOv8 training.

From your description, it seems you've come across an issue where your validation losses are not reflecting expected values, namely, the validation loss and the val/defl loss are stuck at 0, and the val/cls loss is showing "inf". This behavior might be due to several reasons:

1. Empty or Incorrect Validation Data: Make sure your validation set is not empty and that the images and corresponding labels are correctly matched. It's possible that if the validation dataset is empty or not correctly labeled, the loss wouldn't compute as expected.

2. Improper Class Balancing: In cases where some classes are underrepresented, the loss could be dominated by the 'inf' or zeros. Ensure that your dataset has an adequate representation of each class.

3. Learning Rate or Optimization Issues: Sometimes, hyperparameters like the learning rate may not be set optimally, causing the training to not converge during validation. Check your configuration and hyperparameters to ensure they are set appropriately for your dataset.

4. Model Overfitting: If the model fits the training data too closely, it may not perform well on the validation dataset. This could also present itself in the form of unrealistic loss values.

5. Data Normalization: Confirm that the data preprocessing steps applied to the training set are also applied to the validation set. Mismatch in preprocessing can lead to issues during validation loss calculation.

6. Version Mismatch: Ensure that you are using the latest version of YOLOv8, as updates often fix bugs and improve performance.

Without seeing the specific error message or being able to examine your settings in detail, it's challenging to pinpoint the exact issue. However, I recommend re-verifying your validation dataset, double-checking your configuration for any possible errors, and making sure that you're using the latest YOLOv8 version. If the problem persists, consider adjusting the dataset or hyperparameters as necessary.

For detailed guidance on each step and more troubleshooting advice, please refer to our documentation at https://docs.ultralytics.com. If you continue to experience issues, feel free to provide additional information, and the Ultralytics team will be glad to assist you further.

**Comment by starshine360** [comment]
Also, when I trained a yolov8 model for object detection, the val/box_loss was ok, but the val/cls loss was "inf". Is there any suggestion?
![Uploading POPO-20231129-211648.jpg…]()

**Comment by glenn-jocher** [comment]
@starshine360 hi there! If your validation class loss (`val/cls_loss`) is showing "inf", while your validation box loss (`val/box_loss`) appears to be within normal ranges, there are a few potential issues to consider:

1. Label Issues: Check your validation labels for any inaccuracies, particularly regarding class identifiers. Ensure every object in your validation dataset has a correct class label that matches one of the class names defined in your `data.yaml` file.

2. Imbalanced Classes: If certain classes are significantly underrepresented in the training set, the network might not learn those classes well, causing instability in the class loss computation. Look into balancing the classes either by augmenting your dataset or adjusting the class weights.

3. Learning Rates: An improper learning rate can lead to divergences during training, especially for the class loss. Experiment with lowering the learning rate to see if it improves the stability of your training.

4. Model Complexity: Sometimes, a model may be too complex for the amount of data or number of classes you have. Consider simplifying your model or increasing the amount of data for training.

5. Numerical Stability: Ensure that your computing environment is numerically stable. Occasionally, numerical instabilities can cause losses to shoot to infinity, especially if you're using mixed-precision training.

6. Category Encoding: Double-check if there's an issue with the encoding of your classes in the dataset. This can be anything from having a mismatch in the number of classes in your `data.yaml` to improper formatting of the label files.

If you haven't yet, it's worth consulting the Troubleshooting section of our docs. It covers a range of issues and their potential fixes in detail. Additionally, consider reviewing your data preprocessing steps for consistency between your training and validation datasets.

If the problem continues after trying these suggestions, feel free to open an issue on the YOLOv8 repository with detailed logs and configuration files so the team can take a closer look at your training process.

**Comment by starshine360** [comment]
@glenn-jocher Thanks for your advise. Well, when I close the automatic mixed precision (AMP) during the training, it seems to be ok. The result is as below. Are there any side effects of AMP?
![POPO20231130-112207](https://github.com/ultralytics/ultralytics/assets/45736998/6392cdcc-a7be-4d90-8d15-50562d1bd970)

**Comment by glenn-jocher** [comment]
@starshine360 great to hear that disabling Automatic Mixed Precision (AMP) seemed to resolve the issue with the validation class loss!

AMP is a strategy that improves computation speed and reduces memory usage by utilizing half-precision floating-point numbers (float16) for certain operations instead of single-precision (float32). However, using AMP can lead to two main side effects:

1. **Numerical Stability**: The use of float16 can lead to reduced numerical precision, which sometimes might cause issues like underflow or overflow during the computations—this might have been the reason for the "inf" losses you were seeing earlier.

2. **Model Convergence**: In some cases, because of the lower precision, AMP can affect the model's ability to converge to an optimal solution, or it might affect the ultimate accuracy of the model, though this is often not significant.

So, while disabling AMP may result in the loss of the benefits associated with it (like faster training and lower memory consumption), if it adversely affects your model training, it might be the right choice for your specific scenario.

To get the benefits of AMP without these side effects, you can:
- Experiment with a hybrid approach where you use AMP with a larger batch size or adjust the scale of your loss functions.
- Gradually introduce AMP into your training once the loss has stabilized with float32 computation.
- Tweak other hyperparameters, such as learning rate and epsilon values.

If stability issues persist even with these approaches, continue training without AMP. It's better to have a stable training process than a faster, but unreliable one. Always balance speed and stability for your specific task and dataset.

For more information on how AMP affects training and best practices on using it with YOLOv8, check our documentation. Feel free to reach out if you encounter any more concerns. Happy training! 😊👍
